api/app.rb

```diff
@@ -127,11 +128,48 @@ def delete_cloudcmd_cookie
   end
 
   post '/cloudcmd' do
-    # NOTE: The directory *SHOULD* be a relative path as 'cloudcmd' will open it relative
-    # to its root_path (which is the users home directory)
-    dir = params["dir"]
-
     cloudcmd = CloudCmd.new(current_user)
+
+    dir = if params["dir"].present?
+      # Resolve the provided directory into an absolute path, when required
+      # NOTE: expand_path does not modify the path if it is already absolute
+      abs_path = Pathname.new(params["dir"]).expand_path(cloudcmd.root_dir)
+
+      # Attempt to resolve symlinks and any .. sections
+      # NOTE: realpath will error if the file doesn't exist, in this case cleanpath
+      #       is used instead. This is done to prevent the dir query being used to
+      #       poll the file system.
+      abs_path, missing = if abs_path.directory?
+                            [abs_path.realpath, false]
+                          else
+                            [abs_path.cleanpath, true]
+                          end
+
+      # Confirm the path is within cloudcmd's root_dir
+      rel_path = abs_path.relative_path_from(cloudcmd.root_dir).to_s
+      if /\A..\/.*/.match?(rel_path)
```

Comment:
This doesn't look quite right.  Should the first two `.` not be escaped?  Why do we require a `/` after `..`?

Reply:
Notably, if I set `dir=..` I don't get a `422` response.  If I set `dir=../../` I do.

Reply:
Turns out it is actually a bit more complicated because the following is a valid directory:
`/home/centos/..foobar`

Also the regex was just generally broken, fixed now:
https://github.com/openflighthpc/flight-file-manager/pull/12/commits/73332e1949903ee4559406e6ee6464ca05d0c988